net/zerotier/files/etc/config/zerotier

```diff
@@ -1,20 +1,39 @@
 
-config zerotier sample_config
+config zerotier 'global'
+	# Sets whether ZeroTier is enabled or not
 	option enabled 0
-
-	# persistent configuration folder (for ZT controller mode)
+	# Sets the ZeroTier listening port (default 9993; set to 0 for random)
+	#option port '9993'
+	# Client secret (leave blank to generate a secret on first run)
+	option secret ''
+	# Path of the optional file local.conf (see documentation at
+	# https://docs.zerotier.com/config#local-configuration-options)
+	#option local_conf_path '/etc/zerotier.conf'
+	# Persistent configuration directory (to perform other configurations such
+	# as controller mode or moons, etc.)
 	#option config_path '/etc/zerotier'
-	# copy <config_path> to RAM to prevent writing to flash (for ZT controller mode)
+	# Copy the contents of the persistent configuration directory to memory
+	# instead of linking it, this avoids writing to flash
 	#option copy_config_path '1'
 
-	#option port '9993'
-
-	# path to the local.conf
-	#option local_conf '/etc/zerotier.conf'
+# Network configuration, you can have as many configurations as networks you
+# want to join (the network name is optional)
+config network 'mynet'
+	# Identifier of the network you wish to join
+	option id '8056c2e21c000001'
+	# Network configuration parameters (all are optional, if not indicated the
+	# default values are set, see documentation at
+	# https://docs.zerotier.com/config/#network-specific-configuration)
+	option allow_managed '1'
+	option allow_global '0'
+	option allow_default '0'
+	option allow_dns '0'
```

Comment:
If allowing network specific config is the goal of this pull request, I think it's possible to extend current uci config while not breaking backward compatibility.  `join` is still an config option in `zerotier` config section, with network specific configuration in added `network` section.  E.g.

```
config zerotier sample_config
    ...
    list join 8056c2e21c000001

config network 8056c2e21c000001
#	option allow_managed '1'
#	option allow_global '0'
#	option allow_default '0'
#	option allow_dns '0'
```

Reply:
The idea really goes beyond that. What we pursue is what is [commented in the MR](https://github.com/mwarning/zerotier-openwrt/pull/120), to simplify the configuration and the startup script and to eliminate a possibility (the one of starting multiple zerotier daemons) that we know that it does not work correctly and that in addition it does not make too much sense.

Reply:
Actually, the reduction of complexity is for the future and possible new configuration options that may be included in the file. Besides, this configuration is more similar to the one used in other packages, so I understand that being something “more known” it will be less complex for the user.

In any case with the uci-defaults script we have added the configuration migration is covered. I have tested it a few times and it works correctly so there should be no more problems.

Reply:
Multiple Zerotier instances are not useful, since a single instance can connect to multiple networks.
Also the current configuration does not allow any settings on a per network basis. This chance is going to fix this.

Reply:
Current uci config
```
config zerotier sample_config
	option enabled 0
	option config_path '/etc/zerotier'
	option copy_config_path '1'
	option port '9993'
	option local_conf '/etc/zerotier.conf'
	option secret ''

	list join '8056c2e21c000001'
```

Proposed uci config in this pull request (incompatible with current config, requires migration script)
```
config zerotier 'global'
	option enabled 0
	option config_path '/etc/zerotier'
	option copy_config_path '1'
	option port '9993'
	option local_conf_path '/etc/zerotier.conf'
	option secret ''

config network 'mynet'
	option id '8056c2e21c000001'
	option allow_managed '1'
	option allow_global '0'
	option allow_default '0'
	option allow_dns '0'
```

My suggestion is to extend current config like the following.  It maintains compatibility with current uci config.  No loss at all.  Fix the impl so that only one zerotier instance is allowed?  Ok.  Revamp/polish init script code? Ok.  Break the current config?  Don't have to.
```
config zerotier
	option enabled 0
	option config_path '/etc/zerotier'
	option copy_config_path '1'
	option port '9993'
	option local_conf '/etc/zerotier.conf'
	option secret ''

	list join '8056c2e21c000001'

config network '8056c2e21c000001'
	option allow_managed '1'
	option allow_global '0'
	option allow_default '0'
	option allow_dns '0'
```

Reply:
@yousong it is true that with the example that you comment the current configuration would work as it is without changing, but having the migration script that does exactly the same it is preferable to go to the new configuration and leave the package ready for future possibilities.

Let's say that right now we do what you say, we leave a configuration that looks “weird” because I have to declare twice the network I want to join (first a “list join” and then a “network config”). In the next update we would have to change it again and again we would be at “square one”.

I fully understand the initial reluctance of not making the change without the migration script, but now that the `/etc/uci-defaults/80-zt-migration` is already included I don't see the problem.

Reply:
> Let's say that right now we do what you say, we leave a configuration that looks “weird” because I have to declare twice the network I want to join (first a “list join” and then a “network config”). In the next update we would have to change it again and again we would be at “square one”.

Why "In the next update we would have to change it again"?

Looking at it from another angle, I see it as a convenience feature. Only those present in `join` list will be joined.  We can easily join/leave a network by inserting/removing a `#`

Reply:
> Why "In the next update we would have to change it again"?

It is very likely that this will have to be done if new functionality is added. 

> Looking at it from another angle, I see it as a convenience feature. Only those present in `join` list will be joined. We can easily join/leave a network by inserting/removing a `#`

You can do the same by commenting out the only two mandatory lines when configuring a network, e.g:
```
#config network 'mynet'
#	option id '8056c2e21c000001'
```

The advantage of being able to name the networks is that they are easily identifiable by the user.

As I said before, having a migration script that adapts the current configuration to the new one, I don't see the problem.

Reply:
In fact, we have not mentioned it, but one of the objectives we are looking for with this change is precisely the possibility that the user can name their networks and thus be able to identify them much more easily.

Reply:
`name` option can be added to the `network` section if needed.

Well, concerns expressed and still I am not a fan of the incompat change.  The migration script will be a unnecessary burden.  That's it for me at the moment.  Maintainers have the say.